keras_cv/layers/aspp.py

```diff
@@ -0,0 +1,195 @@
+        # to the output.
+        self.aspp_parallel_channels = []
+
+        # Channel1 with Conv2D and 1x1 kernel size.
+        conv_sequential = tf.keras.Sequential(
+            [
+                tf.keras.layers.Conv2D(
+                    filters=self.num_channels, kernel_size=(1, 1), use_bias=False
+                ),
+                tf.keras.layers.BatchNormalization(),
+                tf.keras.layers.Activation(self.activation),
+            ]
+        )
+        self.aspp_parallel_channels.append(conv_sequential)
+
+        # Channel 2 and afterwards are based on self.dilation_rates, and each of them
+        # will have conv2D with 3x3 kernel size.
+        for dilation_rate in self.dilation_rates:
+            conv_sequential = tf.keras.Sequential(
+                [
+                    tf.keras.layers.Conv2D(
+                        filters=self.num_channels,
+                        kernel_size=(3, 3),
+                        padding="same",
+                        dilation_rate=dilation_rate,
+                        use_bias=False,
+                    ),
+                    tf.keras.layers.BatchNormalization(),
+                    tf.keras.layers.Activation(self.activation),
+                ]
+            )
+            self.aspp_parallel_channels.append(conv_sequential)
+
+        # Last channel is the global average pooling with conv2D 1x1 kernel.
+        pool_sequential = tf.keras.Sequential(
+            [
+                tf.keras.layers.GlobalAveragePooling2D(),
+                tf.keras.layers.Reshape((1, 1, channels)),
+                tf.keras.layers.Conv2D(
+                    filters=self.num_channels, kernel_size=(1, 1), use_bias=False
+                ),
+                tf.keras.layers.BatchNormalization(),
+                tf.keras.layers.Activation(self.activation),
+                tf.keras.layers.experimental.preprocessing.Resizing(
+                    height, width, interpolation="bilinear"
+                ),
+            ]
+        )
+        self.aspp_parallel_channels.append(pool_sequential)
+
+        # Final projection layers
+        self.projection = tf.keras.Sequential(
+            [
+                tf.keras.layers.Conv2D(
+                    filters=self.num_channels, kernel_size=(1, 1), use_bias=False
+                ),
+                tf.keras.layers.BatchNormalization(),
+                tf.keras.layers.Activation(self.activation),
+                tf.keras.layers.Dropout(rate=self.dropout),
+            ]
+        )
+
+    def call(self, inputs, training=None):
+        """Calls the Atrous Spatial Pyramid Pooling (ASPP) layer on an input.
+
+        The input the of the ASPP will be a dict of {`level`, `tf.Tensor`}, and ASPP layer
+        will pick the actual input based on the `level` from its init args.
+        The output of ASPP will be a dict of {`level`, `tf.Tensor`} with only one level.
```

Comment:
It depends on the layers that consume the output. Currently the segmentation head will use the level information and infer the scale of the input, so that it can properly scale the output back the same shape. The level serve for that purpose as well.

We could revisit that if you feel a single output tensor is more use friendly.

Reply:
That makes sense to me, as long as the workflow is unified in either dict or single tensor (instead of converting between them in many places), it'd be fine.